### ISSUE TITLE
💅 noRedeclare false positive for overloaded functions in an interface

### ISSUE THREAD

**Issue by jschaf** [biomejs/biome#1932]
### Environment information

```bash
CLI:
  Version:                      1.5.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v21.5.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.2.4"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### Rule name

lint/suspicious/noRedeclare

### Playground link

[Playground](https://biomejs.dev/playground/?lineWidth=120&code=ZQB4AHAAbwByAHQAIABpAG4AdABlAHIAZgBhAGMAZQAgAEUAdgBlAG4AdABNAGEAcAAgAHsACgAgACAAJwBhAHIAdABpAGMAbABlAC0AdgBpAGUAdwBlAGQAJwA6ACAAewAgAGkAZAA6ACAAcwB0AHIAaQBuAGcAIAB9ADsACgAgACAAJwBjAGgAYQB0AC0AcwB0AGEAcgB0AGUAZAAnADoAIAB7ACAAbgBhAG0AZQA6ACAAcwB0AHIAaQBuAGcAOwAgAGUAbQBhAGkAbAA6ACAAcwB0AHIAaQBuAGcAOwAgAHMAdQBiAGoAZQBjAHQAOgAgAHMAdAByAGkAbgBnACAAfQA7AAoAfQAKAAoAZQB4AHAAbwByAHQAIABpAG4AdABlAHIAZgBhAGMAZQAgAFMAbwBtAGUARgB1AG4AYwAgAHsACgAgACAAPABLACAAZQB4AHQAZQBuAGQAcwAgAGsAZQB5AG8AZgAgAEUAdgBlAG4AdABNAGEAcAA%2BACgAYwBtAGQAOgAgACcAbwBuACcALAAgAGUAdgBlAG4AdAA6ACAASwAsACAAYwBiADoAIAAoAGQAYQB0AGEAOgAgAEUAdgBlAG4AdABNAGEAcABbAEsAXQApACAAPQA%2BACAAdgBvAGkAZAApADoAIAB2AG8AaQBkADsACgAgACAAPABLACAAZQB4AHQAZQBuAGQAcwAgAGsAZQB5AG8AZgAgAEUAdgBlAG4AdABNAGEAcAA%2BACgAYwBtAGQAOgAgACcAbwBuAGMAZQAnACwAIABlAHYAZQBuAHQAOgAgAEsALAAgAGMAYgA6ACAAKABkAGEAdABhADoAIABFAHYAZQBuAHQATQBhAHAAWwBLAF0AKQAgAD0APgAgAHYAbwBpAGQAKQA6ACAAdgBvAGkAZAA7AAoAfQAKAA%3D%3D)

### Expected result

Biome should not warn that type parameter `K` is reused because it's part of a different declaration.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by keita-hino** [comment]
I would like to do this, may I?
